Dockerfile

```diff
@@ -17,10 +17,10 @@ ENV PYTHONPATH "/root/.local/lib/python3.12/site-packages"
 COPY requirements requirements
 RUN \
   PIP_CONSTRAINT=requirements/runtime-prerequisites.txt \
-    pip install --user --upgrade --no-cache-dir \
+    pip install --require-hashes --user --upgrade --no-cache-dir \
```

Comment:
This is actually auto-activated when constraint or requirements files have at least one hash. But the problem is that the modern depresolver in pip has a bug that prevents us from using this feature: https://github.com/pypa/pip/issues/9243. I keep hoping I'll get to fixing it upstream one day. You can give it a try if you want, though.